[expo-notifications] Add documentation around push token fetching Promises taking a long time to finish (#10845)
expo/expo: packages/expo-notifications/README.md

# Common gotchas / known issues

### Fetching a push token takes a long time on iOS

> Time it takes `expo-notifications` to fulfill a `Promise` returned by `getDevicePushTokenAsync` or `getExpoPushTokenAsync` may vary and sometimes may be very long on iOS. Citing Apple's [“Troubleshooting Push Notifications” Technical Note](https://developer.apple.com/library/archive/technotes/tn2265/_index.html):

Reviewer: ```suggestion

`getDevicePushTokenAsync` and  `getExpoPushTokenAsync` can sometimes take a long time to resolve on iOS. This is outside of `expo-notifications`'s control, as stated in Apple's [“Troubleshooting Push Notifications” technical note](https://developer.apple.com/library/archive/technotes/tn2265/_index.html):
```

> This is not necessarily an error condition. The system may not have Internet connectivity at all because it is out of range of any cell towers or Wi-Fi access points, or it may be in airplane mode. Instead of treating this as an error, your app should continue normally, disabling only that functionality that relies on push notifications.

> As mentioned, probably the most common reason for this this is a flaky or incompatible Internet connection (fetching a push token requires an Internet connection to register the device in the push notifications service provider) or an invalid configuration of the App ID or Provisioning Profile, but looking at proposed solutions there may be more to it.

Reviewer: ```suggestion
As mentioned, the most common reasons for this this are either an invalid Internet connection (fetching a push token requires an Internet connection to register the device with the service provider) or an invalid configuration of your App ID or Provisioning Profile.
```

> When it comes to solutions historically there are many different ways people have allegedly solved this problem, let's mention at least some of them -- maybe you'll be lucky to solve the problem too!

Reviewer: ```suggestion
Here are a few ways people claim to have solved this problem, maybe one of these will help you solve it, too!
```

- Go read the Apple's [Technical Note on troubleshooting push notifications](https://developer.apple.com/library/archive/technotes/tn2265/_index.html)! This the single most reliable source of information on this problem. To help you grasp what they're suggesting:

>   - Make sure the device has a reliable connection to the Internet (try turning off Wi-Fi or switching to another network, disabling firewall block on port 5223, also suggested in [this SO answer](https://stackoverflow.com/a/34332047/1123156)).

Reviewer: ```suggestion
  - Make sure the device has a reliable connection to the Internet (try turning off Wi-Fi or switching to another network, and disabling firewall block on port 5223, as suggested in [this SO answer](https://stackoverflow.com/a/34332047/1123156)).
```

>   - Make sure your app configuration is set properly for registering for push notifications (for bare workflow check out [this guide](https://developer.apple.com/library/ios/documentation/IDEs/Conceptual/AppDistributionGuide/AddingCapabilities/AddingCapabilities.html#//apple_ref/doc/uid/TP40012582-CH26-SW6), for managed workflow this should be done automatically for you by `expo-cli`) as also suggested by [this StackOverflow answer](https://stackoverflow.com/a/10791240/1123156).

Reviewer: ```suggestion
  - Make sure your app configuration is set properly for registering for push notifications (for bare workflow check out [this guide](https://developer.apple.com/library/ios/documentation/IDEs/Conceptual/AppDistributionGuide/AddingCapabilities/AddingCapabilities.html#//apple_ref/doc/uid/TP40012582-CH26-SW6), for managed workflow this is done automatically for you by `expo-cli`) as also suggested by [this StackOverflow answer](https://stackoverflow.com/a/10791240/1123156).
```

>   - If you're in bare workflow you may also want to try to debug this even further by logging persistent connection debug information as outlined by the note and [this StackOverflow answer](https://stackoverflow.com/a/8036052/1123156) if nothing else works.

Reviewer: ```suggestion
  - If you're in bare workflow you may want to try to debug this even further by logging persistent connection debug information as outlined by [this StackOverflow answer](https://stackoverflow.com/a/8036052/1123156).
```

> - Sometimes APNS servers where the device wants to register may be down as suggested by [this forum thread](https://developer.apple.com/forums/thread/52224). Take a walk and try again!

Reviewer: ```suggestion
- APNS servers near the device may be down as indicated by [this forum thread](https://developer.apple.com/forums/thread/52224). Take a walk and try again later!
```

> **Note:** For Expo's backend to be able to send notifications to your app, you will need to provide it with push notification keys. This can be done using `expo-cli` (`expo credentials:manager`). [Read more in the “Upload notifications credentials” guide](https://docs.expo.io/push-notifications/push-notifications-setup/#credentials).

> > **Note:** Especially on iOS, `Promise` returned by this method may take longer periods of time to fulfill (this method uses [`getDevicePushTokenAsync`](#getdevicepushtokenasync-devicepushtoken) under the hood). For more information see [Fetching a push token takes a long time on iOS](#fetching-a-push-token-takes-a-long-time-on-ios) section of the documentation.

Reviewer: ```suggestion
> **Note:** Especially on iOS, `Promise`s returned by this method may take longer periods of time to fulfill. For more information see [Fetching a push token takes a long time on iOS](#fetching-a-push-token-takes-a-long-time-on-ios).
```

Returns a native APNS, FCM token or a [`PushSubscription` data](https://developer.mozilla.org/en-US/docs/Web/API/PushSubscription) that can be used with another push notification service.

> > **Note:** Especially on iOS, `Promise` returned by this method may take longer periods of time to fulfill. For more information see [Fetching a push token takes a long time on iOS](#fetching-a-push-token-takes-a-long-time-on-ios) section of the documentation.

Reviewer: ```suggestion
> **Note:** Especially on iOS, `Promise`s returned by this method may take longer periods of time to fulfill. For more information see [Fetching a push token takes a long time on iOS](#fetching-a-push-token-takes-a-long-time-on-ios) section of the documentation.
```
